picasso/src/main/java/com/squareup/picasso3/SourceBufferingInputStream.java

```diff
@@ -33,33 +35,49 @@
   private final Buffer buffer;
   private long position;
   private long mark = -1;
+  private long markLimit = -1;
 
   SourceBufferingInputStream(BufferedSource source) {
     this.source = source;
     this.buffer = source.buffer();
   }
 
   private final Buffer temp = new Buffer();
+  // offset is the write offset in the dest array
   private int copyTo(byte[] sink, int offset, int byteCount) {
     // TODO replace this with https://github.com/square/okio/issues/362
-    buffer.copyTo(temp, offset, byteCount);
+    buffer.copyTo(temp, position, byteCount);
     return temp.read(sink, offset, byteCount);
   }
 
   @Override public int read() throws IOException {
-    source.require(position + 1);
+    if (!source.request(position + 1)) {
+      return -1;
+    }
     byte value = buffer.getByte(position++);
-    if (position > mark) {
+    if (position > markLimit) {
       mark = -1;
     }
     return value;
   }
 
   @Override public int read(@NonNull byte[] b, int off, int len) throws IOException {
-    source.require(position + len);
-    int copied = /*buffer.*/copyTo(b, off, len);
+    checkNotNull(b, "b is null");
```

Comment:
`"b == null"`

Reply:
Although I'm not sure we really need this explicit check. `b` is immediately dereferenced by the `if` statement below.